File: app-go/internal/order/domain/order_status.go
// Package domain — Order context
package domain

import (
	"github.com/orkhan/ecommerce/internal/shared/domain"
)

// OrderStatus — state machine
//
// Laravel: OrderStatusEnum.php  ·  Spring: OrderStatusEnum.java
// Go: typed string constants + transition map
type OrderStatus string

const (
	OrderStatusPending   OrderStatus = "PENDING"
	OrderStatusConfirmed OrderStatus = "CONFIRMED"
	OrderStatusPaid      OrderStatus = "PAID"
	OrderStatusShipped   OrderStatus = "SHIPPED"
	OrderStatusDelivered OrderStatus = "DELIVERED"
	OrderStatusCancelled OrderStatus = "CANCELLED"
)

var allowedTransitions = map[OrderStatus]map[OrderStatus]bool{
	OrderStatusPending:   {OrderStatusConfirmed: true, OrderStatusCancelled: true},
	OrderStatusConfirmed: {OrderStatusPaid: true, OrderStatusCancelled: true},
	OrderStatusPaid:      {OrderStatusShipped: true, OrderStatusCancelled: true},
	OrderStatusShipped:   {OrderStatusDelivered: true},
	OrderStatusDelivered: {},
	OrderStatusCancelled: {},
}

func (s OrderStatus) CanTransitionTo(target OrderStatus) bool {
	return allowedTransitions[s][target]
}

func (s OrderStatus) RequireTransitionTo(target OrderStatus) error {
	if !s.CanTransitionTo(target) {
		return domain.NewDomainError("Yanlış status keçidi: " + string(s) + " → " + string(target))
	}
	return nil
}

func (s OrderStatus) IsFinal() bool {
	return s == OrderStatusDelivered || s == OrderStatusCancelled
}

func (s OrderStatus) Label() string {
	switch s {
	case OrderStatusPending:
		return "Gözləyir"
	case OrderStatusConfirmed:
		return "Təsdiqlənib"
	case OrderStatusPaid:
		return "Ödənilib"
	case OrderStatusShipped:
		return "Göndərilib"
	case OrderStatusDelivered:
		return "Çatdırılıb"
	case OrderStatusCancelled:
		return "Ləğv edilib"
	}
	return string(s)
}

File: java/examples/06-order-events/src/main/java/com/example/orders/entity/OrderStatus.java
package com.example.orders.entity;

import java.util.Set;

public enum OrderStatus {
    PENDING, CONFIRMED, SHIPPED, DELIVERED, CANCELLED;

    // State machine: hər statusdan hansı statuslara keçid mümkündür
    public Set<OrderStatus> allowedTransitions() {
        return switch (this) {
            case PENDING   -> Set.of(CONFIRMED, CANCELLED);
            case CONFIRMED -> Set.of(SHIPPED,   CANCELLED);
            case SHIPPED   -> Set.of(DELIVERED);
            case DELIVERED, CANCELLED -> Set.of();
        };
    }

    public void validateTransition(OrderStatus next) {
        if (!allowedTransitions().contains(next)) {
            throw new IllegalStateException(
                    "Yanlış keçid: %s → %s".formatted(this, next));
        }
    }
}
